### PR TITLE
feat(validator/docker_registry): add docker registry validator.

### PR DESCRIPTION
**Pull request by cristhian-castaneda**
Docker Validator added.

Registries tested:
gcr.io
index.docker.io
quay.io
account-id.dkr.ecr.region.amazon.aws.com
server-repo.jfrog.io

### CODE REVIEW

File: pkg/validate/docker_validator.go [hunk 1]
```diff
@@ -0,0 +1,197 @@
...
+	if len(registry.Repositories) != 0 {
+		for _, repository := range registry.Repositories {
+			tagCount, err := service.GetTags(repository)
+
+			if err != nil {
+				return false, err
+			}
+
+			if tagCount == 0 {
+				return false, errors.New(fmt.Sprintf("Repository %s contain any tags. Spinnaker will not be able to deploy any docker images, Push some images to your registry.", repository))
```
Comment (ncknt) [inline pkg/validate/docker_validator.go:166]: Use `fmt.Errorf`. Also errors usually start with lower case in Go.

File: pkg/validate/docker_validator.go [hunk 2]
```diff
@@ -0,0 +1,197 @@
...
+			resolvedPassword = password
+		}
+		if len(resolvedPassword) == 0 || err != nil {
+			return false, errors.New("The supplied password file is empty.")
+		}
+	} else if passwordCommandProvided {
+		out, err := exec.Command("bash", "-c", registry.PasswordCommand).Output()
+
+		if err != nil {
+			return false, errors.New(fmt.Sprintf("password command returned non 0 return code, stderr/stdout was: %s", err))
```
Comment (ncknt) [inline pkg/validate/docker_validator.go:122]: Use fmt.Errorf. Also errors usually start with lower case in Go.

File: pkg/validate/docker_validator.go [hunk 3]
```diff
@@ -0,0 +1,197 @@
...
+				return false, errors.New("Your password file has a trailing newline; many text editors append a newline to files they open." + " If you think this is causing authentication issues, you can strip the newline with the command:\n\n" + " tr -d '\\n' < PASSWORD_FILE | tee PASSWORD_FILE")
+			}
+		}
+
+		return false, errors.New(fmt.Sprintf("Unable to establish a connection with docker registry %s with provided credentials", registry.GetAddress()))
+	}
+
+	if len(registry.Repositories) != 0 {
+		for _, repository := range registry.Repositories {
+			tagCount, err := service.GetTags(repository)
```
Comment (ncknt) [inline pkg/validate/docker_validator.go:159]: Can we do that in goroutines to speed things up?

File: pkg/validate/docker_service.go [hunk 4]
```diff
@@ -0,0 +1,166 @@
...
+
+	if err != nil {
+		return nil, err
+	}
+
+	var resp *http.Response
+	resp, err = s.httpService.Execute(req)
+
+	if err != nil {
+		return nil, err
```
Comment (german-muzquiz) [inline pkg/validate/docker_service.go:65]: Improvement: 
```
return nil, fmt.Errorf("Error making request to %s:\n %w", url, err)
```

File: pkg/validate/docker_service.go [hunk 5]
```diff
@@ -0,0 +1,166 @@
...
+}
+
+func (s *dockerRegistryService) GetBase() (bool, error) {
+	if _, err := s.client("/v2/", nil); err != nil {
+		return false, err
+	}
+	return true, nil
+}
+
+func (s *dockerRegistryService) GetTags(image string) (int, error) {
```
Comment (german-muzquiz) [inline pkg/validate/docker_service.go:27]: Function could be renamed to something like `GetTagsCount`

File: pkg/validate/docker_validator.go (hunk 1 above)
Comment (cristhian-castaneda) [reply pkg/validate/docker_validator.go:166]: Yeah, As this is a message to the user, I'm not sure if we need to follow that convention. What do you think?

File: pkg/validate/docker_validator.go (hunk 3 above)
Comment (cristhian-castaneda) [reply pkg/validate/docker_validator.go:159]: Yeah, I'll do it

File: pkg/util/http.go [hunk 6]
```diff
@@ -0,0 +1,69 @@
...
+	DELETE HttpMethod = "DELETE"
+)
+
+func (s *HttpService) Request(method HttpMethod, url string, requestParams map[string]string, headers map[string]string, body io.Reader) (*http.Request, error) {
+	req, err := http.NewRequest(string(method), url, body)
+	if err != nil {
+		return nil, err
+	}
+
+	req = req.WithContext(context.TODO())
```
[the remark this answers is not in the document]
Comment (cristhian-castaneda) [reply pkg/util/http.go:29]: Yeah, sure, I'll do that.

**Pull request merged**